lib/auth/touchid/api_darwin.go

```diff
@@ -0,0 +1,214 @@
+//go:build touchid
+// +build touchid
+
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package touchid
+
+// #cgo CFLAGS: -Wall -xobjective-c -fblocks -fobjc-arc
+// #cgo LDFLAGS: -framework CoreFoundation -framework Foundation -framework LocalAuthentication -framework Security
+// #include <stdlib.h>
+// #include "authenticate.h"
+// #include "credential_info.h"
+// #include "credentials.h"
+// #include "register.h"
+import "C"
+
+import (
+	"crypto/ecdsa"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"strings"
+	"unsafe"
+
+	"github.com/google/uuid"
+	"github.com/gravitational/trace"
+
+	log "github.com/sirupsen/logrus"
+)
+
+func init() {
+	native = &touchIDImpl{}
+}
+
+type touchIDImpl struct{}
+
+func (*touchIDImpl) IsAvailable() bool {
+	// TODO(codingllama): Write a deeper check that looks at binary
+	//  signature/entitlements/etc.
+	return true
+}
+
+func (*touchIDImpl) Register(rpID, user string, userHandle []byte) (*CredentialInfo, error) {
+	credentialID := uuid.NewString()
+	userHandleB64 := base64.RawURLEncoding.EncodeToString(userHandle)
+
+	var req C.CredentialInfo
+	req.label = C.CString(makeLabel(rpID, user))
+	req.app_label = C.CString(credentialID)
+	req.app_tag = C.CString(userHandleB64)
+	defer func() {
+		C.free(unsafe.Pointer(req.label))
+		C.free(unsafe.Pointer(req.app_label))
+		C.free(unsafe.Pointer(req.app_tag))
+	}()
+
+	var errMsgC, pubKeyC *C.char
+	defer func() {
+		C.free(unsafe.Pointer(errMsgC))
+		C.free(unsafe.Pointer(pubKeyC))
+	}()
+
+	if res := C.Register(req, &pubKeyC, &errMsgC); res != 0 {
+		errMsg := C.GoString(errMsgC)
+		return nil, errors.New(errMsg)
+	}
+
+	pubKeyB64 := C.GoString(pubKeyC)
+	pubKeyRaw, err := base64.StdEncoding.DecodeString(pubKeyB64)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &CredentialInfo{
+		CredentialID: credentialID,
+		publicKeyRaw: pubKeyRaw,
+	}, nil
+}
+
+// rpID are domain names, so it's safe to assume they won't have spaces in them.
+// https://www.w3.org/TR/webauthn-2/#relying-party-identifier
+const labelSeparator = " "
+
+func makeLabel(rpID, user string) string {
+	return fmt.Sprintf("%v%v%v", rpID, labelSeparator, user)
```

Comment:
Simple string concatenation will be more efficient than sprintf and probably just as readable.
```suggestion
	return rpID + labelSeparator + user
```

Reply:
Done.